radixdlt-java/src/main/java/com/radixdlt/client/core/ledger/RadixAtomPuller.java

```diff
@@ -49,6 +54,100 @@ public RadixAtomPuller(
 		this.atomStore = atomStore;
 	}
 
+
+	// HACK
+	// TODO: Move this into a proper reducer framework
+	private static class AtomReducer implements Consumer<RadixNodeAction> {
+		private RadixNode pullNode = null;
+		private final ConcurrentHashMap<RadixHash, Pair<Spin, AtomObservation>> particleSpins = new ConcurrentHashMap<>();
+		private final FetchAtomsAction initialAction;
+		private final ObservableEmitter<AtomObservation> emitter;
+		private final BiConsumer<RadixAddress, AtomObservation> atomStore;
+		private final RadixAddress address;
+
+		AtomReducer(
+			BiConsumer<RadixAddress, AtomObservation> atomStore,
+			FetchAtomsAction initialAction,
+			ObservableEmitter<AtomObservation> emitter
+		) {
+			this.atomStore = atomStore;
+			this.initialAction = initialAction;
+			this.address = initialAction.getAddress();
+			this.emitter = emitter;
+		}
+
+		// TODO: Replace the following checks with constraint machine to
+		// check for conflicts rather than just DOWN particle conflicts
+		private Optional<Atom> getAtomConflict(Atom atom, Particle particle, Spin spin) {
+			if (spin != Spin.DOWN) {
+				return Optional.empty();
+			}
+
+			Pair<Spin, AtomObservation> lastObservation = particleSpins.get(particle.getHash());
+
+			return Optional.ofNullable(lastObservation)
+				.flatMap(o -> !o.getSecond().getAtom().equals(atom) && o.getFirst() == Spin.DOWN
+						? Optional.of(o.getSecond().getAtom())
+						: Optional.empty());
+		}
+
+		@Override
+		public void accept(RadixNodeAction action) {
+			final List<AtomObservation> observations = new ArrayList<>();
+
+			if (action instanceof FetchAtomsObservationAction) {
+
+				FetchAtomsObservationAction fetchAtomsObservationAction = (FetchAtomsObservationAction) action;
+				if (fetchAtomsObservationAction.getUuid().equals(initialAction.getUuid())) {
+					if (pullNode == null) {
+						pullNode = action.getNode();
+					}
+
+					AtomObservation observation = fetchAtomsObservationAction.getObservation();
+					if (observation.hasAtom()) {
+						observation.getAtom().spunParticles()
+							.map(s -> TransitionedParticle.fromSpunParticle(s, observation.getType()))
+							.forEach(t -> {
```

Comment:
`t` is already a `TransitionedParticle` because of the previous `.map(...)`

Reply:
That's what I would expect but for some reason compiler is not being happy on my end when trying to run, it thinks `t` is an `Object`. Any ideas why?

Reply:
Looks like a type inference fail in the Oracle JDK.  Interestingly the eclipse compiler gets this right.  Best leave it as is I guess.